xarray/backends/api.py

```diff
@@ -516,12 +528,20 @@ def open_mfdataset(paths, chunks=None, concat_dim=_CONCAT_DIM_DEFAULT,
     if preprocess is not None:
         datasets = [preprocess(ds) for ds in datasets]
 
-    if concat_dim is _CONCAT_DIM_DEFAULT:
-        combined = auto_combine(datasets, compat=compat)
-    else:
-        combined = auto_combine(datasets, concat_dim=concat_dim, compat=compat)
-    combined._file_obj = _MultiFileCloser(file_objs)
-    combined.attrs = datasets[0].attrs
+    # close datasets in case of a ValueError
+    try:
+        if concat_dim is _CONCAT_DIM_DEFAULT:
+            combined = auto_combine(datasets, compat=compat,
+                                    data_vars=data_vars)
+        else:
+            combined = auto_combine(datasets, concat_dim=concat_dim,
+                                    compat=compat, data_vars=data_vars)
+        combined._file_obj = _MultiFileCloser(file_objs)
+        combined.attrs = datasets[0].attrs
+    except ValueError as ve:
```

Comment:
Let's only wrap the lines where this could fail -- so this should be moved up two lines, before `combined._file_obj` is assigned.